### PR TITLE
Set `mkl_set_dynamic` to false in `init_num_threads`

### LINKED ISSUE #90760: Large slow down by not calling `torch.set_num_threads`
### 🐛 Describe the bug

I am running a Python program in a computer with 20 cores. By default PyTorch sets the number of threads to the number of cores. In this computer, if I don't call `torch.set_num_threads`, the script below takes 7 seconds, but if I call it the script takes 4 seconds. Note that in both cases the script is using the same number of threads.

In another computer with 6 cores, if I do not call `set_num_threads` the script takes 5.3 seconds, but if I call it the script takes 4.98 seconds.

The script below is a minimal working example. I am experiencing very large slow downs, on the order of 10, on a more complex piece of code.

This [discussion](https://discuss.pytorch.org/t/how-to-check-if-pytorch-is-using-blas/50068/3) pointed me in the right direction to understand this problem.

``` python                                                                                                                                                                                                                                             
import sys
import time
import argparse
import torch

def main(argv):
    parser = argparse.ArgumentParser()
    parser.add_argument("--set_num_threads", action="store_true",
                        help="set the number of threads to the number of cores")
    parser.add_argument("--n_iter", type=int, help="number of iterations", default=100)
    parser.add_argument("--n_rows", type=int, help="number of rows",
                        default=1000)
    parser.add_argument("--n_cols", type=int, help="number of columns",
                        default=1000)
    args = parser.parse_args()

    set_num_threads = args.set_num_threads
    n_iter = args.n_iter
    n_rows = args.n_rows
    n_cols = args.n_cols

    if set_num_threads > 0:
        print("called torch.set_num_threads")
        torch.set_num_threads(torch.get_num_threads())
    else:
        print("NOT called torch.set_num_threads")

    print(f"Using {torch.get_num_threads()} threads")

    start_time = time.time()
    for i in range(n_iter):
        m = torch.randn((n_rows, n_cols), dtype=torch.double)
        b = torch.randn((n_rows, 1), dtype=torch.double)
        m.requires_grad = True
        L = torch.linalg.cholesky(torch.matmul(m, m.T))
        x = torch.cholesky_solve(b, L)
        m.requires_grad = False
    end_time = time.time()
    print(f"Processing time {end_time-start_time}")

    breakpoint()


if __name__ == "__main__":
    main(sys.argv)
```

### Versions

Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.17

Python version: 3.7.13 (default, Mar 29 2022, 02:18:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-132-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] gpytorch==1.2.1
[pip3] mypy==0.910
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.5
[pip3] numpydoc==1.4.0
[pip3] torch==1.12.1
[pip3] torchvision==0.13.1a0
[conda] _pytorch_select           0.2                       gpu_0  
[conda] blas                      1.0                         mkl  
[conda] cpuonly                   1.0                           0    pytorch
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] gpytorch                  1.2.1                    pypi_0    pypi
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py37h7f8727e_0  
[conda] mkl_fft                   1.3.1            py37hd3c417c_0  
[conda] mkl_random                1.2.2            py37h51133e4_0  
[conda] numpy                     1.21.5           py37h6c91a56_3  
[conda] numpy-base                1.21.5           py37ha15fc14_3  
[conda] numpydoc                  1.4.0            py37h06a4308_0  
[conda] pytorch                   1.12.1          cpu_py37he8d8e81_0  
[conda] torchvision               0.13.1          cpu_py37h164cc8f_0

### PR DESCRIPTION
**Pull request by sanchitintel**
Dynamic adjustment of the number of threads by MKL is enabled by default.
So, MKL may use fewer threads in parallel regions than the number returned by `mkl_get_max_threads`.
With libgomp (GCC's OpenMP), this may result in the creation & destruction of threads across every
MKL & non-MKL boundary of OpenMP usage, resulting in degraded performance.
So, it should be disabled by default by calling `set_mkl_dynamic(false)`. `set_num_threads` in `ParallelOpenMP.cpp` disables it, anyway.

As a follow-up, maybe we can add an environment variable, which true, 
would not call `set_mkl_dynamic(false)`, as doing so would yield better performance with Intel OpenMP,
so users who would like to use Intel OpenMP would benefit.

Fixes #90760

cc @jgong5 @mingfeima @XiaobingSuper @ashokei @jingxu10

### CODE REVIEW

**Comment by sanchitintel** [comment]
@jgong5, looks like we don't need to call [`mkl_set_dynamic(false)` in `set_num_threads`](https://github.com/pytorch/pytorch/blob/02fe86eb49b71827e19527ef0998165c785a4e8f/aten/src/ATen/ParallelOpenMP.cpp#L65) if we would call it in `init_num_threads` now.

File: aten/src/ATen/ParallelOpenMP.cpp [hunk 1]
```diff
@@ -37,6 +37,11 @@ void init_num_threads() {
     // size of the OpenMP thread pool, resulting in worse performance (and memory
     // leaks in GCC 5.4)
     omp_set_num_threads(mkl_get_max_threads());
+    // Dynamic adjustment of the number of threads by MKL is enabled by default.
+    // So, MKL may use fewer threads in parallel regions than the number returned by mkl_get_max_threads.
+    // With libgomp (GCC's OpenMP), this may result in the creation & destruction of threads across every
+    // MKL & non-MKL boundary of OpenMP usage, resulting in degraded performance.
```
Comment (jgong5) [inline aten/src/ATen/ParallelOpenMP.cpp:43]: I don't understand why there would be "creation & destruction" cost if MKL uses less number of OMP threads. OMP runtime should maintain a pool of threads and don't have to "create/destroy" threads for every OMP region?

**Review by jgong5 — COMMENTED** [review]
I feel this is a big change that requires extensive benchmarking to make sure no perf regression happens.

**Comment by sanchitintel** [comment]
> I feel this is a big change that requires extensive benchmarking to make sure no perf regression happens.

I agree. I'll benchmark with TorchBench. Thanks!

**Comment by ashokei** [comment]
@sanchitintel did you verify https://github.com/pytorch/pytorch/issues/90760 is resolved with this fix?

**Comment by sanchitintel** [comment]
Hi @jgong5 @aemani, while using `mkl_set_dynamic(false)` would fix #90760 by preventing the destruction & re-creation of OpenMP thread-pools at each MKL & non-MKL boundary, doing so may result in MKL using more threads than it'd otherwise have used with the default setting (`MKL_DYNAMIC` being true), which, in turn, [may lead to a numerical accuracy issue (higher divergence in output than what PyTorch's tolerance bounds allow) due to different ordering of floating-point  operations](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-linux-developer-guide/top/obtaining-numerically-reproducible-results.html).

eg. with the current master branch, the existing unit-test 
`MKL_DYNAMIC=FALSE PYTORCH_TEST_WITH_SLOW=1 MKL_VERBOSE=1 python test/test_jit.py -v TestModels.test_neural_style` would fail because its output diverges more than what `checkTrace`'s tolerance bounds allows (this issue was exposed in this PR's CI failures).

However, since `set_mkl_dynamic(false)` is being called in the current master branch with `torch.set_num_threads()` invocations anyway, can we go ahead with this change, viz. using `set_mkl_dynamic(false)` by default as well, if there are few possible instances of divergence in outputs exceeding tolerance bounds?  If not, it looks like we can't fix #90760 due to PyTorch's dependency on libgomp, but can only offer preloading Intel OpenMP as a workaround. Please weigh in. Thanks!
